java/arcs/core/storage/handle/CollectionImpl.kt

```diff
@@ -90,6 +95,11 @@ class CollectionImpl<T : Referencable>(
         log.debug { "Storing: $entity" }
         @Suppress("GoodTime") // use Instant
         entity.creationTimestamp = requireNotNull(time).currentTimeMillis
+        if (entity is RawEntity && !schema.refinement(entity)) {
+            throw IllegalArgumentException(
+                "Invalid entity stored to handle $name (failed refinement)"
+            )
+        }
```

Comment:
Consider this equivalent alternative:

```kotlin
require(entity !is RawEntity || schema.refinement(entity)) {
    "Invalid entity stored to handle $name (failed refinement)"
}
```

Reply:
+1

Aside: It bothers me a little that this is `!is` rather than `is!`